pkg/kubevirt/client.go

```diff
@@ -270,64 +291,85 @@ func (c *client) CreateVolumeSnapshot(ctx context.Context, namespace, name, clai
+	if util.Contains(snapshotClassNames, snapshotClassName) {
+		return snapshotClassName, nil
+	}
+	if !(c.storageClassEnforcement.AllowAll || c.storageClassEnforcement.AllowDefault) {
+		tenantSnapshotClasses := c.getTenantSnapshotClassesFromInfraStorageClassName(storageClassName)
+		if len(tenantSnapshotClasses) > 0 {
+			if snapshotClassName == "" {
+				return "", fmt.Errorf("unable to determine volume snapshot class name for snapshot creation, valid snapshot classes are %v", tenantSnapshotClasses)
+			} else {
+				return "", fmt.Errorf("volume snapshot class %s is not compatible with PVC with storage class %s, valid snapshot classes for this pvc are %v", snapshotClassName, storageClassName, tenantSnapshotClasses)
+			}
+		} else {
+			return "", fmt.Errorf("unable to determine volume snapshot class name for snapshot creation, no valid snapshot classes found")
+		}
 	}
-	snapshotClass, err := c.getSnapshotClassFromStorageClass(ctx, storageClassName, snapshotClassName)
-	if err != nil {
-		return "", err
+	return "", nil
+}
+
+func (c *client) getInfraSnapshotClassesFromStorageClassName(storageClassName string) []string {
+	for _, storageSnapshotMapping := range c.infraTenantStorageSnapshotMapping {
+		infraStorageClasses := []string{}
+		infraSnapshotClasses := []string{}
+		for _, storageClasses := range storageSnapshotMapping.StorageClasses {
+			infraStorageClasses = append(infraStorageClasses, storageClasses.Infra)
+		}
+		for _, snapshotClasses := range storageSnapshotMapping.VolumeSnapshotClasses {
+			infraSnapshotClasses = append(infraSnapshotClasses, snapshotClasses.Infra)
+		}
+		if util.Contains(infraStorageClasses, storageClassName) {
+			return infraSnapshotClasses
+		}
 	}
-	return snapshotClass.Name, nil
+	return nil
 }
 
-func (c *client) isStorageClassAllowed(ctx context.Context, storageClassName string) (bool, error) {
-	if !c.storageClassEnforcement.AllowAll && !util.Contains(c.storageClassEnforcement.AllowList, storageClassName) {
-		if c.storageClassEnforcement.AllowDefault {
-			// Check if storage class is default and default is allowed.
-			storageClass, err := c.kubernetesClient.StorageV1().StorageClasses().Get(ctx, storageClassName, metav1.GetOptions{})
-			if err != nil {
-				if errors.IsNotFound(err) {
-					return false, nil
-				} else {
-					return false, err
-				}
-			}
-			ann := storageClass.GetAnnotations()
-			if v, ok := ann["storageclass.kubernetes.io/is-default-class"]; !ok || v != "true" {
-				return false, nil
-			}
-		} else {
-			return false, nil
+func (c *client) getTenantSnapshotClassesFromInfraStorageClassName(storageClassName string) []string {
+	for _, storageSnapshotMapping := range c.infraTenantStorageSnapshotMapping {
+		infraStorageClasses := []string{}
+		tenantSnapshotClasses := []string{}
+		for _, storageClasses := range storageSnapshotMapping.StorageClasses {
+			infraStorageClasses = append(infraStorageClasses, storageClasses.Infra)
+		}
```

Comment:
you could use a similar pattern to the one i recommended for getInfraSnapshotClassesFromStorageClassName here

Reply:
Done